templates/tfengine/components/app/cicd/main.tf

```diff
@@ -0,0 +1,194 @@
+  for_each = toset(var.build_viewers)
+  project  = data.google_project.devops.project_id
+  role     = "roles/viewer"
+  member   = each.value
+}
+
+# IAM permissions to allow Cloud Build Service Account use the billing account.
+resource "google_billing_account_iam_member" "binding" {
+  billing_account_id = local.constants.billing_account
+  role               = "roles/billing.user"
+  member             = local.cloudbuild_sa
+}
+
+# Cloud Build - Cloud Build Service Account IAM permissions
+# IAM permissions to allow Cloud Build SA to access state.
+resource "google_storage_bucket_iam_member" "cloudbuild_state_iam" {
+  bucket = local.constants.state_bucket
+  role   = "roles/storage.admin"
+  member = local.cloudbuild_sa
+}
+
+locals {
+  folder_ids = [for env in module.constants.values: env.folder_id]
+  folder_roles = flatten([
+    for f in local.folder_ids : [
+      for r in local.cloudbuild_sa_editor_roles : {
+        folder = f
+        role   = r
+      }
+    ]
+  ])
+}
+
+# Grant Cloud Build Service Account access to the folder.
+resource "google_folder_iam_member" "cloudbuild_sa_folder_iam" {
+  for_each = { for fr in local.folder_roles : "${fr.folder} ${fr.role}" => fr }
+  folder   = each.value.folder
+  role     = each.value.role
+  member   = local.cloudbuild_sa
+}
+
+resource "google_sourcerepo_repository" "configs" {
+  project  = data.google_project.devops.project_id
+  name     = var.source_repo_name
+}
+
+# Grant Cloud Build Service Account access to the devops project.
+resource "google_project_iam_member" "cloudbuild_sa_project_iam" {
+  for_each = toset(local.cloudbuild_devops_roles)
+  project  = data.google_project.devops.project_id
+  role     = each.key
+  member   = local.cloudbuild_sa
+}
+
+resource "google_cloudbuild_trigger" "validate" {
+  provider    = google-beta
+  project     = data.google_project.devops.project_id
+  name        = "tf-validate"
+  description = "Terraform validate job triggered on push event."
+
+  included_files = [
+    "${local.terraform_root_prefix}**",
+  ]
+  trigger_template {
+    repo_name   = google_sourcerepo_repository.configs.name
+    branch_name = var.branch_name
+  }
+  filename = "${local.terraform_root_prefix}cicd/configs/tf-validate.yaml"
+  substitutions = {
+    _TERRAFORM_ROOT = local.terraform_root
+  }
+}
+
+resource "google_cloudbuild_trigger" "plan" {
+  provider    = google-beta
+  project     = data.google_project.devops.project_id
+  name        = "tf-plan"
+  description = "Terraform plan job triggered on push event."
+
+  included_files = [
+    "${local.terraform_root_prefix}**",
+  ]
+  trigger_template {
+    repo_name   = google_sourcerepo_repository.configs.name
+    branch_name = var.branch_name
+  }
+  filename = "${local.terraform_root_prefix}cicd/configs/tf-plan.yaml"
+  substitutions = {
+    _TERRAFORM_ROOT = local.terraform_root
+  }
+}
+
+resource "google_cloudbuild_trigger" "apply" {
```

Comment:
Am I correct in understanding that this trigger will always be created, but will fail every time unless the right permissions are granted to the devops SA to apply changes?

Reply:
The permissions are granted on line 119

Reply:
I meant that if folder_roles doesn't have permissions to apply, this trigger will fail, right?

Reply:
Oh right. Well the deployment itself will fail

Reply:
Is there a reasonable way to disable this trigger if the permissions aren't there?

Reply:
I can't think of one. If the user does not have permission to grant the SA access on the folder then it should give a pretty clear warning. If there is a permission on the SA missing then when the trigger runs it will highlight the problem there, so I think we are covered in most cases.